packages/flutter/lib/src/material/page_transitions_theme.dart

```diff
@@ -176,133 +178,162 @@ class _ZoomPageTransition extends StatefulWidget {
-  AnimationStatus _lastAnimationStatus;
+class _ZoomedTransitions extends StatelessWidget {
+  const _ZoomedTransitions({
+    @required this.animation,
+    @required this.child,
+  });
+
+  final Animation<double> animation;
+  final Widget child;
 
   @override
-  void initState() {
-    super.initState();
-    widget.animation.addStatusListener((AnimationStatus animationStatus) {
-      _lastAnimationStatus = _currentAnimationStatus;
-      _currentAnimationStatus = animationStatus;
-    });
+  Widget build(BuildContext context) {
+    return DualTransitionBuilder(
+      animation: animation,
+      forwardBuilder: (
+        BuildContext context,
+        Animation<double> animation,
+        Widget child,
+      ) {
+        return _ZoomPageTransitionIn(
+          animation: animation,
+          child: child,
+        );
+      },
+      reverseBuilder: (
+        BuildContext context,
+        Animation<double> animation,
+        Widget child,
+      ) {
+        return _ZoomPageTransitionOut(
+          animation: ReverseAnimation(animation),
+          child: child,
+        );
+      },
+      child: child,
+    );
   }
+}
 
-  // This check ensures that the animation reverses the original animation if
-  // the transition were interruped midway. This prevents a disjointed
-  // experience since the reverse animation uses different fade and scaling
-  // curves.
-  bool get _transitionWasInterrupted {
-    bool wasInProgress = false;
-    bool isInProgress = false;
-
-    switch (_currentAnimationStatus) {
-      case AnimationStatus.completed:
-      case AnimationStatus.dismissed:
-        isInProgress = false;
-        break;
-      case AnimationStatus.forward:
-      case AnimationStatus.reverse:
-        isInProgress = true;
-        break;
-    }
-    switch (_lastAnimationStatus) {
-      case AnimationStatus.completed:
-      case AnimationStatus.dismissed:
-        wasInProgress = false;
-        break;
-      case AnimationStatus.forward:
-      case AnimationStatus.reverse:
-        wasInProgress = true;
-        break;
-    }
-    return wasInProgress && isInProgress;
+class _ZoomPageTransitionIn extends StatelessWidget {
+  const _ZoomPageTransitionIn({
+    this.child,
+    this.animation,
+  });
+
+  final Widget child;
+  final Animation<double> animation;
+
+  @override
+  Widget build(BuildContext context) {
+    // Only change color when the animation was completed.
+    final bool isCompleted = animation.status == AnimationStatus.completed;
+    return Container(
+      color: Colors.black.withOpacity(
+        (isCompleted
+            ? kAlwaysDismissedAnimation
+            : _ZoomPageTransition._forwardScrimOpacityTween.animate(animation)
+        ).value,
+      ),
+      child: ScaleTransition(
+        scale: (
+          !isCompleted
+              ? _ZoomPageTransition._forwardEndScreenScaleTween
+              : _ZoomPageTransition._reverseStartScreenScaleTween
```

Comment:
Why this? It appears that both tweens produce the same value when the animation is completed?

Reply:
Yes but it produce different value when animating. It require different value between the forward end and the reverse start, as the implementation before.